### PR TITLE
Fix shuffle history (2)

### LINKED ISSUE #4108: Shuffle mode loses played-history when song list is reordered
Steps to reproduce
------------------

1. Enable random shuffle
2. Start a song (_song A_)
3. Press "next" button
4. Sort song list by a different column so the order of the songs change
5. Press "previous" button (a second time if playback goes to the beginning of the current song)

Expected Output
---------------

Playback goes to _song A_.

Actual Output
-------------

Playback stops. UI shows "Not playing".

Test System
-----------

Quod Libet 4.5.0

Additional Information
----------------------

Because of this issue auto-sorting was disabled while shuffle mode is active in https://github.com/quodlibet/quodlibet/pull/3581. The auto-sorting feature seems to have made this bug more obvious, but it's not the root cause. This also happens when the list is manually re-sorted.

Also related: #3592

### PR DESCRIPTION
**Pull request by slosd**
Check-list
----------

 * [x] There is a linked issue discussing the motivations for this feature or bugfix
 * [x] Unit tests have been added where possible
 * [ ] I've added / updated documentation for any user-facing features.
 * [x] Performance seems to be comparable or better than current `main`

What this change is adding / fixing
-----------------------------------

This is https://github.com/quodlibet/quodlibet/pull/4255 again, with a fix for the usage of invalid iters (#4373 and #4381).

For reference I looked at the super class `TrackCurrentModel` which keeps an iter around in `__iter`. My implementation does not hold on to iters for longer than that class.

Fixes #4108

### CODE REVIEW

File: quodlibet/order/__init__.py [hunk 1]
```diff
@@ -103,39 +105,42 @@ class OrderRemembered(Order):
...
 
-    def remaining(self, playlist):
+    def remaining(self, playlist) -> Dict[int, Any]:
         """Gets a map of all song indices to their song from the `playlist`
         that haven't yet been played"""
-        all_indices = set(range(len(playlist)))
-        played = set(self._played)
+
+        def get_index(iter):
+            return playlist.get_path(iter).get_indices()[0]
```
Comment (declension) [inline quodlibet/order/__init__.py:136]: I remember this line coming up a bit whilst debugging. Can we put some null-saftety (etc) here?

If somehow `playlists.get_path(iter)` returns `None`. Or `get_indices()` does, or is empty?...
If they're at least on separate lines, stack traces are a bit quicker to read

**Review by declension — CHANGES_REQUESTED** [review]
Thanks. I think it looks better but haven't had time to understand fully.

Ideally I'd like someone on Arch to test this so we can be sure the crashes won't happen.

File: quodlibet/order/__init__.py (hunk 1 above)
Comment (slosd) [reply quodlibet/order/__init__.py:136]: I'll split up the expression into multiple lines. Other than that, I'm worried that not throwing on `None` or an empty list will make it harder to discover incorrect code.

**Comment by declension** [comment]
Thanks

**Pull request merged**